keras/layers/wrappers.py

```diff
@@ -152,9 +157,33 @@ def compute_output_shape(self, input_shape):
         timesteps = input_shape[1]
         return (child_output_shape[0], timesteps) + child_output_shape[1:]
 
+    def _reshape_input_mask(self, mask):
+        if mask is not None and K.ndim(mask) > 2:
+            mask_shape = K.shape(mask)
+            mask_shape = tuple(mask_shape[i] for i in range(K.ndim(mask)))
+            return K.reshape(mask, (-1,) + mask_shape[2:])
+        return None
+
+    def compute_mask(self, inputs, mask):
+        input_shape = K.int_shape(inputs)
+        input_length = input_shape[1]
+        if input_length is None:
+            input_length = K.shape(inputs)[1]
+        inputs = K.reshape(inputs, (-1,) + input_shape[2:])
+        reshaped_mask = self._reshape_input_mask(mask)
+        output_mask = self.layer.compute_mask(inputs, reshaped_mask)
+        if output_mask is not None:
+            output_mask_shape = K.shape(output_mask)
+            output_mask_shape = tuple(output_mask_shape[i] for i in range(K.ndim(output_mask)))
+            return K.reshape(output_mask, (-1, input_length) + output_mask_shape[1:])
+        elif mask is not None and K.ndim(mask) > 2:
+            return K.any(mask, range(K.ndim(mask))[2:])
+        else:
+            return mask
+
     def call(self, inputs, mask=None):
         input_shape = K.int_shape(inputs)
-        if input_shape[0]:
+        if input_shape[0] and (mask is None or K.ndim(mask) <= 2):
```

Comment:
1. I always have this question,  why batch size matters there? Does reshape-method fail in this case?
2. The behaviours of `K.rnn` with mask is NOT consistent with those using reshaped, as `K.rnn` carries on previous value when current time slice is masked.